File: api/services/embeddings.py
"""Embedding service using Cohere."""
import cohere
from api.config import settings
from typing import List


_cohere_client = None


def get_cohere_client():
    """Get Cohere client instance."""
    global _cohere_client
    if _cohere_client is None:
        _cohere_client = cohere.Client(api_key=settings.cohere_api_key)
    return _cohere_client


def create_embedding(text: str) -> List[float]:
    """
    Create an embedding for the given text.
    
    Args:
        text: Text to embed
    
    Returns:
        List of float values representing the embedding vector
    """
    client = get_cohere_client()
    response = client.embed(
        texts=[text],
        model="embed-english-v3.0",
        input_type="search_document"
    )
    return response.embeddings[0]

File: frontend/app/lib/services/embeddings.ts
/** Embedding service using Cohere. */
import { CohereClient } from 'cohere-ai';
import { getEnv } from '../utils';

let cohereClient: CohereClient | null = null;

/**
 * Get Cohere client instance.
 */
function getCohereClient(): CohereClient {
  if (cohereClient === null) {
    const apiKey = getEnv('COHERE_API_KEY');
    if (!apiKey) {
      throw new Error('COHERE_API_KEY environment variable is required');
    }
    cohereClient = new CohereClient({ token: apiKey });
  }
  return cohereClient;
}

/**
 * Create an embedding for the given text.
 */
export async function createEmbedding(text: string): Promise<number[]> {
  const client = getCohereClient();
  
  const response = await client.embed({
    texts: [text],
    model: 'embed-english-v3.0',
    inputType: 'search_document',
  });
  
  return (response.embeddings as number[][])[0];
}

/**
 * Create embeddings for multiple texts.
 */
export async function createEmbeddings(texts: string[]): Promise<number[][]> {
  const client = getCohereClient();
  
  const response = await client.embed({
    texts,
    model: 'embed-english-v3.0',
    inputType: 'search_document',
  });
  
  return response.embeddings as number[][];
}

/**
 * Create a query embedding (uses different input type for search queries).
 */
export async function createQueryEmbedding(text: string): Promise<number[]> {
  const client = getCohereClient();
  
  const response = await client.embed({
    texts: [text],
    model: 'embed-english-v3.0',
    inputType: 'search_query',
  });
  
  return (response.embeddings as number[][])[0];
}

